modules/platforms/cpp/common/ignite_result.h

```diff
@@ -0,0 +1,337 @@
+ */
+
+#pragma once
+
+#include <cstdint>
+#include <string>
+#include <optional>
+#include <future>
+#include <functional>
+
+#include "common/ignite_error.h"
+
+namespace ignite
+{
+
+/**
+ * Ignite Result.
+ */
+template<typename T>
+class IgniteResult
+{
+public:
+    // Default
+    IgniteResult() = default;
+    ~IgniteResult() = default;
+    IgniteResult(IgniteResult&&) noexcept = default;
+    IgniteResult(const IgniteResult&) = default;
+    IgniteResult& operator=(IgniteResult&&) noexcept = default;
+    IgniteResult& operator=(const IgniteResult&) = default;
+
+    /**
+     * Constructor.
+     *
+     * @param value Value.
+     */
+    IgniteResult(T&& value) : // NOLINT(google-explicit-constructor)
+        m_value(std::forward<T>(value)),
+        m_error(std::nullopt) { }
+
+    /**
+     * Constructor.
+     *
+     * @param message Message.
+     */
+    IgniteResult(IgniteError&& error) : // NOLINT(google-explicit-constructor)
+        m_value(std::nullopt),
+        m_error(std::forward<IgniteError>(error)) { }
+
+    /**
+     * Has value.
+     *
+     * @return @c true if the result has value.
+     */
+    [[nodiscard]]
+    bool hasValue() const noexcept
+    {
+        return m_value.has_value();
+    }
+
+    /**
+     * Has error.
+     *
+     * @return @c true if the result has error.
+     */
+    [[nodiscard]]
+    bool hasError() const noexcept
+    {
+        return m_error.has_value();
+    }
+
+    /**
+     * Get value.
+     *
+     * @return Value.
+     */
+    [[nodiscard]]
+    T&& getValue()
+    {
+        if (!hasValue())
+            throw IgniteError("No value is present in result");
+
+        return std::move(m_value.value());
+    }
+
+    /**
+     * Get value.
+     *
+     * @return Value.
+     */
+    [[nodiscard]]
+    const T& getValue() const
+    {
+        if (!hasValue())
+            throw IgniteError("No value is present in result");
+
+        return m_value.value();
+    }
+
+    /**
+     * Get error.
+     *
+     * @return Error.
+     */
+    [[nodiscard]]
+    IgniteError&& getError()
+    {
+        if (!hasError())
+            throw IgniteError("No error is present in result");
+
+        return std::move(m_error.value());
+    }
+
+    /**
+     * Get error.
+     *
+     * @return Error.
+     */
+    [[nodiscard]]
+    const IgniteError& getError() const
+    {
+        if (!hasError())
+            throw IgniteError("No error is present in result");
+
+        return m_error.value();
+    }
+
+    /**
+     * Bool operator. Can be used to check result for an error.
+     *
+     * @return @c true if result does not contain error.
+     */
+    explicit operator bool() const noexcept {
+        return !hasError();
+    }
+
+    /**
+     * Wrap operation result in IgniteResult.
+     * @param operation Operation to wrap.
+     * @return IgniteResult
+     */
+    static IgniteResult ofOperation(std::function<T()>&& operation) noexcept {
```

Comment:
I guess, you are right. This will also make it not necessary to specify a type of result value. However, let's move it to `common` lib once we refactor it properly. I've left a TODO.